File: sds/features.py
from __future__ import annotations

import re
from dataclasses import dataclass

from .models import ExtractedFeature


@dataclass(frozen=True)
class FeatureVector:
    credible_eyewitnesses: float
    identification_quality: float
    forensic_dna: float
    forensic_fingerprint: float
    weapon_recovered: float
    confession_recorded: float
    cctv_present: float
    chain_of_custody: float
    statement_quality: float
    expert_forensic: float
    disclosure_complete: float
    hearsay_risk: float
    time_to_arrest_score: float

    def as_dict(self) -> dict[str, float]:
        return {
            "credible_eyewitnesses": self.credible_eyewitnesses,
            "identification_quality": self.identification_quality,
            "forensic_dna": self.forensic_dna,
            "forensic_fingerprint": self.forensic_fingerprint,
            "weapon_recovered": self.weapon_recovered,
            "confession_recorded": self.confession_recorded,
            "cctv_present": self.cctv_present,
            "chain_of_custody": self.chain_of_custody,
            "statement_quality": self.statement_quality,
            "expert_forensic": self.expert_forensic,
            "disclosure_complete": self.disclosure_complete,
            "hearsay_risk": self.hearsay_risk,
            "time_to_arrest_score": self.time_to_arrest_score,
        }

    def as_list(self) -> list[float]:
        return list(self.as_dict().values())


FEATURE_NAMES = list(FeatureVector(0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0).as_dict().keys())


def _count(patterns: tuple[str, ...], text: str) -> int:
    return sum(1 for p in patterns if re.search(p, text, re.I))


def _bool(patterns: tuple[str, ...], text: str) -> float:
    return 1.0 if any(re.search(p, text, re.I) for p in patterns) else 0.0


def extract_features(text: str) -> tuple[FeatureVector, list[ExtractedFeature]]:
    """NLP + rule-based feature extraction from unstructured police narratives."""
    lower = text.lower()
    details: list[ExtractedFeature] = []

    witness_terms = (
        r"\b(independent witness|eyewitness|credible witness|corroborated witness)\b",
        r"\b(multiple witnesses|two witnesses|three witnesses)\b",
    )
    witness_hits = _count(witness_terms, lower)
    witness_count = min(3.0, witness_hits + (1.0 if re.search(r"\bwitness\b", lower) else 0.0))
    details.append(
        ExtractedFeature(
            "credible_eyewitnesses",
            witness_count,
            f"{witness_count:.0f} eyewitness signal(s)",
            "nlp",
        )
    )

    id_good = _bool(
        (
            r"\b(identification parade|line[- ]?up|CCTV (still|footage)|Turnbull)\b",
            r"\b(independent witness|multiple witnesses)\b",
        ),
        lower,
    )
    id_bad = _bool(
        (
            r"\b(show[- ]?up|single photo|brief glimpse|poor lighting|one witness only)\b",
        ),
        lower,
    )
    identification_quality = max(0.0, min(1.0, id_good - 0.5 * id_bad))
    details.append(
        ExtractedFeature(
            "identification_quality",
            identification_quality,
            "Identification procedure quality (0–1)",
            "nlp",
        )
    )

    forensic_dna = _bool((r"\b(DNA|genetic profile|STR profile)\b",), lower)
    forensic_fingerprint = _bool((r"\b(fingerprint|latent print|AFIS)\b",), lower)
    weapon_recovered = _bool((r"\b(firearm recovered|weapon seized|knife recovered|gun found)\b",), lower)
    confession_recorded = _bool(
        (r"\b(confession|admission|recorded interview|caution administered)\b",),
        lower,
    )
    cctv_present = _bool((r"\b(CCTV|footage|DVR|video clip)\b",), lower)

    chain_good = _bool(
        (r"\b(chain of custody|sealed bag|custody log|property room)\b",),
        lower,
    )
    chain_bad = _bool((r"\b(gap in custody|unsealed|broken seal|no continuity)\b",), lower)
    chain_of_custody = max(0.0, min(1.0, chain_good - 0.6 * chain_bad))

    stmt_good = _bool(
        (r"\b(recorded interview|signed statement|contemporaneous notes)\b",),
        lower,
    )
    stmt_bad = _bool((r"\b(verbal only|unsigned|not recorded|no caution)\b",), lower)
    statement_quality = max(0.0, min(1.0, stmt_good - 0.6 * stmt_bad))

    expert_forensic = _bool(
        (r"\b(expert report|forensic examiner|laboratory report|ISO 17025)\b",),
        lower,
    )
    disclosure_complete = _bool(
        (r"\b(disclosure schedule|unused material|prosecution bundle)\b",),
        lower,
    )
    hearsay_risk = _bool(
        (r"\b(hearsay|someone told me|rumour|second[- ]hand)\b",),
        lower,
    )

    time_match = re.search(
        r"\b(within (\d+) (hours?|days?)|(\d+) hours? (later|after)|same day arrest)\b",
        lower,
    )
    time_to_arrest_score = 0.5
    if time_match:
        time_to_arrest_score = 0.9 if "hour" in time_match.group(0) or "same day" in time_match.group(0) else 0.7
    if re.search(r"\b(months? later|years? later|delayed arrest)\b", lower):
        time_to_arrest_score = 0.2

    for name, value, label in (
        ("forensic_dna", forensic_dna, "DNA / genetic evidence"),
        ("forensic_fingerprint", forensic_fingerprint, "Fingerprint evidence"),
        ("weapon_recovered", weapon_recovered, "Weapon recovered"),
        ("confession_recorded", confession_recorded, "Recorded confession / interview"),
        ("cctv_present", cctv_present, "CCTV / video evidence"),
        ("chain_of_custody", chain_of_custody, "Chain of custody integrity"),
        ("statement_quality", statement_quality, "Witness statement quality"),
        ("expert_forensic", expert_forensic, "Expert / lab report"),
        ("disclosure_complete", disclosure_complete, "Disclosure schedule present"),
        ("hearsay_risk", hearsay_risk, "Hearsay risk indicator"),
        ("time_to_arrest_score", time_to_arrest_score, "Timeliness of arrest (proxy)"),
    ):
        details.append(ExtractedFeature(name, value, label, "nlp"))

    vector = FeatureVector(
        credible_eyewitnesses=witness_count,
        identification_quality=identification_quality,
        forensic_dna=forensic_dna,
        forensic_fingerprint=forensic_fingerprint,
        weapon_recovered=weapon_recovered,
        confession_recorded=confession_recorded,
        cctv_present=cctv_present,
        chain_of_custody=chain_of_custody,
        statement_quality=statement_quality,
        expert_forensic=expert_forensic,
        disclosure_complete=disclosure_complete,
        hearsay_risk=hearsay_risk,
        time_to_arrest_score=time_to_arrest_score,
    )
    return vector, details

File: web/src/lib/features.ts
import type { ExtractedFeature, FeatureVector } from "./types";

function count(patterns: RegExp[], text: string): number {
  return patterns.filter((p) => p.test(text)).length;
}

function bool(patterns: RegExp[], text: string): number {
  return patterns.some((p) => p.test(text)) ? 1 : 0;
}

const p = (source: string) => new RegExp(source, "i");

export function extractFeatures(text: string): {
  vector: FeatureVector;
  details: ExtractedFeature[];
} {
  const lower = text.toLowerCase();
  const details: ExtractedFeature[] = [];

  const witnessHits = count(
    [
      p(String.raw`\b(independent witness|eyewitness|credible witness|corroborated witness)\b`),
      p(String.raw`\b(multiple witnesses|two witnesses|three witnesses)\b`),
    ],
    lower,
  );
  const credibleEyewitnesses = Math.min(3, witnessHits + (/\bwitness\b/i.test(lower) ? 1 : 0));
  details.push({
    name: "credibleEyewitnesses",
    value: credibleEyewitnesses,
    label: `${credibleEyewitnesses} eyewitness signal(s)`,
    source: "nlp",
  });

  const idGood = bool(
    [
      p(String.raw`\b(identification parade|line[- ]?up|CCTV (still|footage)|Turnbull)\b`),
      p(String.raw`\b(independent witness|multiple witnesses)\b`),
    ],
    lower,
  );
  const idBad = bool(
    [p(String.raw`\b(show[- ]?up|single photo|brief glimpse|poor lighting|one witness only)\b`)],
    lower,
  );
  const identificationQuality = Math.max(0, Math.min(1, idGood - 0.5 * idBad));
  details.push({
    name: "identificationQuality",
    value: identificationQuality,
    label: "Identification procedure quality (0–1)",
    source: "nlp",
  });

  const forensicDna = bool([p(String.raw`\b(DNA|genetic profile|STR profile)\b`)], lower);
  const forensicFingerprint = bool([p(String.raw`\b(fingerprint|latent print|AFIS)\b`)], lower);
  const weaponRecovered = bool(
    [p(String.raw`\b(firearm recovered|weapon seized|knife recovered|gun found)\b`)],
    lower,
  );
  const confessionRecorded = bool(
    [p(String.raw`\b(confession|admission|recorded interview|caution administered)\b`)],
    lower,
  );
  const cctvPresent = bool([p(String.raw`\b(CCTV|footage|DVR|video clip)\b`)], lower);

  const chainGood = bool(
    [p(String.raw`\b(chain of custody|sealed bag|custody log|property room)\b`)],
    lower,
  );
  const chainBad = bool([p(String.raw`\b(gap in custody|unsealed|broken seal|no continuity)\b`)], lower);
  const chainOfCustody = Math.max(0, Math.min(1, chainGood - 0.6 * chainBad));

  const stmtGood = bool(
    [p(String.raw`\b(recorded interview|signed statement|contemporaneous notes)\b`)],
    lower,
  );
  const stmtBad = bool([p(String.raw`\b(verbal only|unsigned|not recorded|no caution)\b`)], lower);
  const statementQuality = Math.max(0, Math.min(1, stmtGood - 0.6 * stmtBad));

  const expertForensic = bool(
    [p(String.raw`\b(expert report|forensic examiner|laboratory report|ISO 17025)\b`)],
    lower,
  );
  const disclosureComplete = bool(
    [p(String.raw`\b(disclosure schedule|unused material|prosecution bundle)\b`)],
    lower,
  );
  const hearsayRisk = bool(
    [p(String.raw`\b(hearsay|someone told me|rumour|second[- ]hand)\b`)],
    lower,
  );

  let timeToArrestScore = 0.5;
  if (/\b(within \d+ hours?|same day arrest)\b/i.test(lower)) timeToArrestScore = 0.9;
  else if (/\b(within \d+ days?)\b/i.test(lower)) timeToArrestScore = 0.7;
  if (/\b(months? later|years? later|delayed arrest)\b/i.test(lower)) timeToArrestScore = 0.2;

  const vector: FeatureVector = {
    credibleEyewitnesses,
    identificationQuality,
    forensicDna,
    forensicFingerprint,
    weaponRecovered,
    confessionRecorded,
    cctvPresent,
    chainOfCustody,
    statementQuality,
    expertForensic,
    disclosureComplete,
    hearsayRisk,
    timeToArrestScore,
  };

  for (const [name, value, label] of [
    ["forensicDna", forensicDna, "DNA / genetic evidence"],
    ["forensicFingerprint", forensicFingerprint, "Fingerprint evidence"],
    ["weaponRecovered", weaponRecovered, "Weapon recovered"],
    ["confessionRecorded", confessionRecorded, "Recorded confession / interview"],
    ["cctvPresent", cctvPresent, "CCTV / video evidence"],
    ["chainOfCustody", chainOfCustody, "Chain of custody integrity"],
    ["statementQuality", statementQuality, "Witness statement quality"],
    ["expertForensic", expertForensic, "Expert / lab report"],
    ["disclosureComplete", disclosureComplete, "Disclosure schedule present"],
    ["hearsayRisk", hearsayRisk, "Hearsay risk indicator"],
    ["timeToArrestScore", timeToArrestScore, "Timeliness of arrest (proxy)"],
  ] as const) {
    details.push({ name, value, label, source: "nlp" });
  }

  return { vector, details };
}

export function vectorToList(v: FeatureVector): number[] {
  return [
    v.credibleEyewitnesses,
    v.identificationQuality,
    v.forensicDna,
    v.forensicFingerprint,
    v.weaponRecovered,
    v.confessionRecorded,
    v.cctvPresent,
    v.chainOfCustody,
    v.statementQuality,
    v.expertForensic,
    v.disclosureComplete,
    v.hearsayRisk,
    v.timeToArrestScore,
  ];
}
